### DISCUSSION TITLE
Chain cache always looking for item till the last chain

### DISCUSSION THREAD

**Discussion by AtCliffUnderline** [symfony/symfony#53825] [Q&A] [answered]
### Symfony version(s) affected

6.4.2

### Description

In my example there is `ChainAdapter` with `ArrayAdapter` at first and `RedisTagAwareAdapter` at last. As you can see from profile, when i am trying to retrieve item - chain always goes to RedisTagAwareAdapter and making Redis call. However, item being saved in ArrayAdapter, so I dont understand why Redis is called

### How to reproduce

Code to reproduce, but you should replace with your own client.
```
$cache = new TagAwareAdapter(
            new ChainAdapter(
                [
                    new ArrayAdapter(300, false, 300),
                    new RedisAdapter(
                        $this->redisConnectionPoolService->getClientByPoolName('cache'),
                        'redis.tag.aware.symfony6',
                        86400,
                        new DefaultMarshaller(false),
                    ),
                ],
                300,
            ),
        );

        for ($i = 0; $i < 10; $i++) {
            $item = $cache->getItem('test');
            if (! $item->isHit()) {
                $item->set('value');
                $item->expiresAfter(3600);
                $item->tag(['tag']);
            }
            $cache->save($item);
        }
```
And profile screenshot from KCacheGrind:
![image](https://github.com/symfony/symfony/assets/41080058/f006150b-40f9-4754-a890-5713b9b742e6)

**Comment by xabbuh** [comment]
The `ArrayAdapter` keeps items in the memory. Meaning that it will lose them all when the PHP process terminates. So as long as you don’t have a long running process where you query the same item more than once the behaviour you describe looks expected to me.

**Comment by ivanfilatov** [comment]
@xabbuh Hi, sorry to doubt your message, but from my point of view the idea of `ArrayAdapter` is to keep data in runtime memory not to query them throughout even the single http request, isn't it? What is the point of storing some data in "global" memory for one-time query at all? I can put it in the local variable instead.
More to say, we are now heading to long-living application (using Roadrunner), where some data is worth storing in runtime between requests rather than quering it from Redis for every request. And for long-living case this situation is even more sensitive for us.

**Comment by xabbuh** [comment]
I missed that you query for the same item inside the loop. Can you create a small example application that allows to reproduce your issue?

**Comment by nicolas-grekas** [comment]
> `$cache->save($item);`

This line goes to all adapters so that all of them have it. That's the explanation why redis is always hit.

**Comment by AtCliffUnderline** [comment]
> This line goes to all adapters so that all of them have it.

Well, this is pretty understandable, but I still don't think it is okay to call Redis fetch 10 times.

**Comment by nicolas-grekas** [comment]
You need to move that line inside the "if". If you don't want that redis to be called 10 times, just don't write the code that asks for it :) 
Closing as there is no bug here.

**Comment by nicolas-grekas** [comment]
Argh, I'm sorry, I double checked to be sure and the reason is more complex!
What happens is that the key that is fetch at every iteration is not "test" but "\x01tags\x01test".
Let me understand how this happens and get back to you.

**Reply by AtCliffUnderline** [reply]
Really good news, many thanks!

**Reply by nicolas-grekas** [reply]
git bisect tells me this was introduced in https://github.com/symfony/symfony/pull/42997/commits/6a90ca0fef83a0bb27839b461a5f2396a93d90f6

**Reply by nicolas-grekas** [reply, marked as answer]
Should be fixed by #53846

**Reply by AtCliffUnderline** [reply]
It is awesome, thank you! May I already check it?

**Reply by nicolas-grekas** [reply]
of course

**Reply by AtCliffUnderline** [reply]
seems like it works very well. Before and after on my production code.
![image](https://github.com/symfony/symfony/assets/41080058/8599ff1d-dea8-49df-9c7a-3c2e4abf8883)
![image](https://github.com/symfony/symfony/assets/41080058/f4dda704-db26-4134-9377-71d812d0896a)

**Reply by AtCliffUnderline** [reply]
Looking forward to this patch in release :)
Thanks again!
